analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -94,61 +95,67 @@ public override void ExecutionCompleted()
-        var secondaryLocations = SecondaryLocations(block, conditionValue);
+        // For SwitchExpressionArms like `true => 5` we are only interested in the left part (`true`).
+        var syntax = operation.Syntax.IsKind(SyntaxKindEx.SwitchExpressionArm) ? ((SwitchExpressionArmSyntaxWrapper)operation.Syntax).Pattern : operation.Syntax;
+        var secondaryLocations = SecondaryLocations(block, conditionValue, syntax);
         if (secondaryLocations.Any())
         {
-            ReportIssue(Rule2583, operation, secondaryLocations, issueMessage, S2583MessageSuffix);
+            ReportIssue(Rule2583, syntax, secondaryLocations, issueMessage, S2583MessageSuffix);
         }
         else
         {
-            ReportIssue(Rule2589, operation, null, issueMessage, string.Empty);
+            ReportIssue(Rule2589, syntax, null, issueMessage, string.Empty);
         }
     }
 
-    private List<Location> SecondaryLocations(BasicBlock block, bool conditionValue)
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionValue, SyntaxNode conditionSyntax)
     {
         List<Location> locations = new();
-        IOperation currentStart = null;
-        IOperation currentEnd = null;
-        var unreachable = UnreachableOperations(block, conditionValue).Where(x => !IsIgnoredLocation(x.Syntax)).ToHashSet();
-        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        var unreachable = UnreachableOperations(block, conditionValue);
+        var currentStart = conditionSyntax.Span.End;
+        var reachedNodes = reachedOperations.Select(x => x.Syntax).Where(x => x.SpanStart > conditionSyntax.Span.End).OrderBy(x => x.SpanStart);
+
+        foreach (var node in reachedNodes)
         {
-            if (unreachable.Contains(operation))
+            if (AddUnreachableLocation(node.SpanStart))
             {
-                currentStart ??= operation;
-                currentEnd = operation;
-            }
-            else
-            {
-                AddCurrent();
+                currentStart = node.Span.End;
             }
         }
-        AddCurrent();
+        AddUnreachableLocation(int.MaxValue);  // Get all unreachable operations from the very last reached one until the end of the method.
         return locations;
 
-        void AddCurrent()
+        bool AddUnreachableLocation(int spanMax)
         {
-            if (currentStart is not null)
+            var nodes = unreachable.Where(x => x.SpanStart > currentStart && x.Span.End < spanMax);
+            if (nodes.Any())
             {
-                locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
-                currentStart = null;
+                var firstNode = nodes.OrderBy(x => x.SpanStart).First();
+                var lastNode = nodes.OrderBy(x => x.Span.End).Last();
+                locations.Add(firstNode.CreateLocation(lastNode));
+                return true;
```

Comment:
The double orderby is expensive. You can do this:
   var nodes = unreachable.Where(x => x.SpanStart > currentStart && x.Span.End < spanMax).OrderBy(x => x.SpanStart).ThenBy(x => x.Span.End);
            if (nodes.Length <> 0)
            {
                locations.Add(nodes.First().CreateLocation(nodes.Last()));
                return true;

Reply:
Since operations can overlap, this changes the location for the worse.